allennlp/tests/training/gan_callback_trainer_test.py

```diff
@@ -0,0 +1,439 @@
+        )
+    else:
+        return _original_optimizer_from_params(model_parameters, params)
+
+# This is also bad, I have to provide a do-nothing dataset reader.
+@DatasetReader.register("nil")
+class NilDatasetReader(DatasetReader):
+    # pylint: disable=abstract-method
+    def _read(self, file_path):
+        return [Instance({})]
+
+
+def config(sample_size: int = 500,
+           batches_per_epoch: int = 40,
+           num_epochs: int = 50,
+           learning_rate: float = 0.05) -> Params:
+    return Params({
+        # These are unnecessary but the TrainerPieces.from_params expects them
+        "iterator": {"type": "basic"},
+        "dataset_reader": {"type": "nil"},
+        "train_data_path": "",
+
+        # These are necessary
+        "model": {
+            "type": "gan",
+            "generator": {
+                "type": "generator2-test",
+                "input_dim": 1,
+                "hidden_dim": 5,
+                "output_dim": 1
+            },
+            "discriminator": {
+                "type": "discriminator2-test",
+                "input_dim": sample_size,
+                "hidden_dim": 5,
+                "preprocessing": "moments"
+            }
+        },
+        "trainer": {
+            "type": "callback",
+            "optimizer": {
+                "type": "gan",
+                "generator_optimizer": {
+                    "type": "sgd",
+                    "lr": learning_rate
+                },
+                "discriminator_optimizer": {
+                    "type": "sgd",
+                    "lr": learning_rate
+                }
+            },
+            "num_epochs": num_epochs,
+            "callbacks": [
+                {
+                    "type": "train-gan"
+                },
+                {
+                    "type": "generate-gan-training-batches",
+                    "batch_size": sample_size,
+                    "batches_per_epoch": batches_per_epoch,
+                    "sampler": {
+                        "type": "normal",
+                        "mean": 4.0,
+                        "stdev": 1.25
+                    },
+                    "noise_sampler": {
+                        "type": "uniform"
+                    }
+                }
+            ]
+
+        }
+    })
+
+
+class GanCallbackTrainerTest(AllenNlpTestCase):
+    def setUp(self):
+        super().setUp()
+        MonkeyPatch().setattr(Optimizer, 'from_params', _optimizer_from_params)
+
+    def test_gan_can_train(self):
+        params = config(batches_per_epoch=2, num_epochs=2)
+        train_model(params, self.TEST_DIR)
+
+
+if __name__ == "__main__":
+    # Run it yourself, it's fun!
+    #
+    # python -m allennlp.tests.training.gan_callback_trainer_test
+    #
+    # pylint: disable=invalid-name
+    MonkeyPatch().setattr(Optimizer, 'from_params', _optimizer_from_params)
+
+
+    from allennlp.training.trainer_base import TrainerBase
+    serialization_dir_ = tempfile.mkdtemp()
```

Comment:
Why the underscores?  Because of pylint?  But you already disabled `invalid-name`...